core/modules/field/modules/options/options.module

```diff
@@ -382,3 +386,289 @@ function options_field_widget_error($element, $error, $form, &$form_state) {
+    '#default_value_allowed' => TRUE,
+    '#default_value_pattern' => '',
+    '#element_validate' => array('form_options_validate'),
+    '#theme_wrappers' => array('form_element'),
+  );
+
+  return $type;
+}
+
+/**
+ * Expand the "options" form element type.
+ *
+ * The "options" type is simply an enhanced textarea that makes it easier to
+ * create key|value pairs and put items into optgroups.
+ */
+function form_options_expand($element) {
+  module_load_include('inc', 'options', 'options.element');
+  return _form_options_expand($element);
+}
+
+/**
+ * Validate the "options" form element type.
+ */
+function form_options_validate($element, &$form_state) {
+  module_load_include('inc', 'options', 'options.element');
+  _form_options_validate($element, $form_state);
+}
+
+/**
+ * This function adjusts the value of the element from a text value to an array.
+ */
+function form_type_options_value(&$element, $edit = FALSE) {
+  module_load_include('inc', 'options', 'options.element');
+  return _form_type_options_value($element, $edit);
+}
+
+/**
+ * Create a textual representation of options from an array.
+ *
+ * @param $options
+ *   An array of options used in a select list.
+ * @param $key_type
+ *   How key/value pairs should be interpreted. Available options:
+ *   - mixed
+ *   - numeric
+ *   - associative
+ *   - custom
+ *   - none
+ */
+function form_options_to_text($options, $key_type) {
+  module_load_include('inc', 'options', 'options.element');
+  return _form_options_to_text($options, $key_type);
+}
+
+/**
+ * Create an array representation of text option values.
+ *
+ * If the Key of the option is within < >, treat as an optgroup
+ *
+ * <Group 1>
+ *   creates an optgroup with the label "Group 1"
+ *
+ * <>
+ *   Exits the current group, allowing items to be inserted at the root element.
+ */
+function form_options_from_text($text, $key_type, $flat = FALSE, &$duplicates = array()) {
+  module_load_include('inc', 'options', 'options.element');
+  return _form_options_from_text($text, $key_type, $flat, $duplicates);
+}
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function options_form_field_ui_field_settings_form_alter(&$form, &$form_state) {
+  $type = $form['field']['type']['#value'];
+
+  if (in_array($type, _options_element_supported_field_types())) {
+    // Get default value.
+    $options = form_options_from_text($form['field']['settings']['allowed_values']['#default_value'], 'mixed');
+    // Toggled by default if options already set.
+    $key_type_toggled = !empty($options);
+
+    $element_overwrite = array(
+      '#options' => $options,
+      '#multiple' => FALSE,
+      '#key_type_toggled' => $key_type_toggled,
+      '#default_value_allowed' => FALSE,
+    );
+    _options_element_add_allowed_values_element($form, $element_overwrite);
+  }
+}
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function options_form_field_ui_field_edit_form_alter(&$form, &$form_state) {
```

Comment:
We should not need to alter our own core modules. The existing form definition in `options_field_widget_form()` should just be updated directly.

Reply:
This is a bit tricky. It seems that the original options_elements alters the instance and field UI forms and not just the widget. I think it was replacing the textarea in "global settings" (when on edit form) whereas the widget is not in there (and not on the settings page?).

I've edited `options_field_widget_settings_form` and it appears on the edit page for the field. But if I edit `options_field_widget_form` it will also appear on the front end.

Reply:
I tried `hook_field_settings_form()` but for some reason it's not getting called. It seems to be right direction. Though I wonder why options_elements didn't use these hooks in the first place?

Reply:
I think the point is to move the code in this `_alter` directly into the form it is altering, which would be `field_ui_field_edit_form`. I dont quite understand why you'd need to modify `hook_field_settings_form() and hook_field_instance_settings_form()`

Reply:
I don't think it can be `field_ui_field_edit_form` since options could be disabled. I'm now thinking it should be in `list` module since `list_field_instance_settings_form`, `list_field_settings_form` are being called and the list and options modules are a bit intertwined.

Reply:
I'm confused. `options_form_field_ui_field_edit_form_alter()` alters a form with FORM ID `field_ui_field_edit_form()` (its an implementation of `hook_form_FORM_ID_alter()`). There is no other place that this code should go.

Reply:
Yeah, I can confirm that it works in `list_field_settings_form`. Needs a bit more work but basically works like before. And I suppose that means that options is a dependency list if it isn't already.

Reply:
field_ui_field_edit_form calls a hook, hook_field_settings_form which is where it should go IMO. That's how these modules seem to already doing things.